lnwallet/fee_estimator.go

```diff
@@ -524,3 +549,206 @@ func (s SparseConfFeeSource) ParseResponse(r io.Reader) (map[uint32]uint32, erro
+}
+
+// RelayFeePerKW returns the minimum fee rate required for transactions to be
+// relayed.
+//
+// NOTE: This method is part of the FeeEstimator interface.
+func (w *WebAPIFeeEstimator) RelayFeePerKW() SatPerKWeight {
+	return FeePerKwFloor
+}
+
+// randomFeeUpdateTimeout returns a random timeout between MinFeeUpdateTimeout
+// and MaxFeeUpdateTimeout that will be used to determine when the Estimator
+// should retrieve fresh fees from its API.
+func (w *WebAPIFeeEstimator) randomFeeUpdateTimeout() time.Duration {
+	lower := int64(MinFeeUpdateTimeout)
+	upper := int64(MaxFeeUpdateTimeout)
+	return time.Duration(prand.Int63n(upper-lower) + lower)
+}
+
+// getCachedFee takes in a target for the number of blocks until an initial
+// confirmation and returns an estimated fee (if one was returned by the API). If
+// the fee was not previously cached, we cache it here. Note that the feesMtx
+// should be held while calling this function, to avoid a race condition.
+func (w *WebAPIFeeEstimator) getCachedFee(numBlocks uint32) (uint32, error) {
+	w.feesMtx.Lock()
+	defer w.feesMtx.Unlock()
+	for target := numBlocks; target >= MinBlockTarget; target-- {
+		fee, ok := w.feeByBlockTarget[target]
+		if !ok {
+			continue
+		}
+
+		w.feeByBlockTarget[numBlocks] = fee
+		return fee, nil
+	}
+	return 0, fmt.Errorf("web API does not include a fee estimation for "+
+		"block target of %v", numBlocks)
+}
+
+// updateFeeEstimates re-queries the API for fresh fees and caches them.
+func (w *WebAPIFeeEstimator) updateFeeEstimates() {
+	// Rather than use the default http.Client, we'll make a custom one
+	// which will allow us to control how long we'll wait to read the
+	// response from the service. This way, if the service is down or
+	// overloaded, we can exit early and use our default fee.
+	netTransport := &http.Transport{
+		Dial: (&net.Dialer{
+			Timeout: 5 * time.Second,
+		}).Dial,
+		TLSHandshakeTimeout: 5 * time.Second,
+	}
+	netClient := &http.Client{
+		Timeout:   time.Second * 10,
+		Transport: netTransport,
+	}
+
+	// With the client created, we'll query the API source to fetch the URL
+	// that we should use to query for the fee estimation.
+	targetURL := w.apiSource.GenQueryURL()
+	resp, err := netClient.Get(targetURL)
+	if err != nil {
+		walletLog.Errorf("unable to query web api for fee response: %v",
+			err)
+		return
+	}
+	defer resp.Body.Close()
+
+	// Once we've obtained the response, we'll instruct the WebAPIFeeSource
+	// to parse out the body to obtain our final result.
+	feesByBlockTarget, err := w.apiSource.ParseResponse(resp.Body)
+	if err != nil {
+		walletLog.Errorf("unable to query web api for fee response: %v",
+			err)
+		return
+	}
+
+	w.feesMtx.Lock()
+	w.feeByBlockTarget = feesByBlockTarget
+	w.feesMtx.Unlock()
+}
+
+// feeUpdateManager updates the fee estimates whenever a new block comes in.
+func (w *WebAPIFeeEstimator) feeUpdateManager() {
+	defer w.wg.Done()
+
+	for {
+		select {
+		case <-w.updateFeeTimer.C:
```

Comment:
could we just update the fee estimate on demand instead of having a goroutine running to keep it up to date?

Reply:
Hm, what do you think about updating it on demand, but only if X minutes have passed since it was last updated? Could be worthwhile to avoid the API getting overwhelmed?

Reply:
Yeah, idea is to request the fee and cache it first time the estimator is queried, then request it again only after some time has passed. Would that lead to more requests to the API than the current approach?

Reply:
The timer approach means that the load on w/e API is bounded and can be extrapolated ahead of time (based on the number of clients). Otherwise, (within lnd) routine things like sweeping multiple HTLCs or updating the commitment fee rate will trigger far more load if we query directly each time we need an estimate.

Reply:
It would go like this:
```go
func EstimateFee(blocks int) {
    _, ok := cache[blocks]
    if time.Now() > lastTime.Add(cacheInvalidationTimeout) || !ok {
          fetchAndAddToCache(...)
    }

    return cache[blocks]
```
}